File: 04/tyrion.py
import sys

n = int(sys.stdin.readline())
house_index = dict()
last_index = 0
clauses = []

for i in range(n):
    line = sys.stdin.readline().split()
    parity = 1 if line[-1] == 'odd' else 0
    houses = line[:-1]

    row = []
    for house in houses:
        if house not in house_index:
            house_index[house] = last_index
            last_index += 1
        row.append(house_index[house])
    clauses.append((row, parity))

A = [[0]*last_index for i in range(len(clauses))]
b = [0]*len(clauses)

for i, (row, parity) in enumerate(clauses):
    for idx in row:
        A[i][idx] = 1
    b[i] = parity


# Gaussian elimination over GF(2). Assume linear system is solvable
def gauss(A, b):
    n = len(A)
    m = len(A[0])
    row = 0
    for col in range(m):
        pivot = -1
        for i in range(row, n):
            if A[i][col]:  # A[i][col] == 1
                pivot = i
                break
        if pivot == -1:
            continue
        A[row], A[pivot] = A[pivot], A[row]  # swap row and pivot
        b[row], b[pivot] = b[pivot], b[row]  # swap row and pivot
        for i in range(n):
            if i != row and A[i][col]:
                for j in range(m):
                    A[i][j] ^= A[row][j]  # A[i][j] = A[i][j] - A[row][j]
                b[i] ^= b[row]  # b[i] = b[i] - b[row]
        row += 1
    return b


solution = gauss(A, b)
print(sum(solution))

File: 04/Tyrion.java
import java.util.*;
import java.util.stream.IntStream;

public class Tyrion {

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int n = Integer.parseInt(input.nextLine());

        Map<String, Integer> houseIndex = new HashMap<>();
        List<BitSet> A = new ArrayList<>();
        List<Integer> b = new ArrayList<>();
        int varCount = 0; //keep track of house index

        for (int i = 0; i < n; i++) {
            String[] parts = input.nextLine().trim().split("\\s+");
            int parity = parts[parts.length - 1].equals("odd") ? 1 : 0;

            BitSet row = new BitSet();
            for (int j = 0; j < parts.length - 1; j++) {
                String house = parts[j];
                if (houseIndex.putIfAbsent(house, varCount) == null) varCount++;
                row.set(houseIndex.get(house));
            }
            A.add(row);
            b.add(parity);
        }
        int m = varCount;
        int rows = A.size();
        BitSet[] matrix = A.toArray(new BitSet[0]);
        int[] rhs = b.stream().mapToInt(i -> i).toArray();

        int row = 0;
        for (int col = 0; col < m; col++) {
            int pivot = -1;
            for (int i = row; i < rows; i++) {
                if (matrix[i].get(col)) {
                    pivot = i;
                    break;
                }
            }
            if (pivot == -1) {
                continue;
            }

            // swap row and pivot
            BitSet tmp = matrix[row];
            matrix[row] = matrix[pivot];
            matrix[pivot] = tmp;
            int temp = rhs[row];
            rhs[row] = rhs[pivot];
            rhs[pivot] = temp;

            for (int i = 0; i < rows; i++) {
                if (i != row && matrix[i].get(col)) {
                    matrix[i].xor(matrix[row]);
                    rhs[i] ^= rhs[row];
                }
            }
            row++;
        }

        int solution = IntStream.of(rhs).sum();
        System.out.println(IntStream.of(solution).sum());
    }
}
